src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs

```diff
@@ -359,17 +389,25 @@ private static void RemoveDirectoryInternal(string fullPath, bool topLevel, bool
 
         public static void SetAttributes(string fullPath, FileAttributes attributes)
         {
-            if (!Interop.Kernel32.SetFileAttributes(fullPath, (int)attributes))
+            if (Interop.Kernel32.SetFileAttributes(fullPath, (int)attributes))
             {
-                int errorCode = Marshal.GetLastWin32Error();
-                if (errorCode == Interop.Errors.ERROR_INVALID_PARAMETER)
-                    throw new ArgumentException(SR.Arg_InvalidFileAttrs, nameof(attributes));
-                throw Win32Marshal.GetExceptionForWin32Error(errorCode, fullPath);
+                return;
             }
+
+            int errorCode = Marshal.GetLastWin32Error();
+            if (errorCode == Interop.Errors.ERROR_INVALID_PARAMETER)
+                throw new ArgumentException(SR.Arg_InvalidFileAttrs, nameof(attributes));
+            throw Win32Marshal.GetExceptionForWin32Error(errorCode, fullPath);
         }
 
-        // Default values indicate "no change".  Use defaults so that we don't force callsites to be aware of the default values
-        private static unsafe void SetFileTime(
+        public static void SetAttributes(SafeFileHandle fileHandle, FileAttributes attributes)
+        {
+            ThrowHelper.ThrowForMissingPath_SafeFileHandle(fileHandle.Path);
+            SetAttributes(fileHandle.Path!, attributes);
+        }
```

Comment:
This API should use `SetFileInformationByHandle` and I think it shouldn't throw for a null `fileHandle.Path`, instead, let the pinvoke decide if something is wrong with the handle.

You could use `SetFileAttributesW` when `fileHandle.Path` is not null but only if SetFileAttributesW shows better performance than SetFileInformationByHandle i.e: as an optimization.

Reply:
Is there a special documentation to benchmark WinAPI Methods / Do benchmarks on the file system, not have a large deviation?

Reply:
> Do benchmarks on the file system, not have a large deviation?

The benchmarks we have in place are valid and a good indicator for regressions. In other words, the deviation is not that large. 

> Is there a special documentation to benchmark WinAPI Methods

I would use BenchmarkDotNet https://benchmarkdotnet.org/articles/guides/getting-started.html to compare the two methods.

Reply:
@Jozkee `SetFileInformationByHandle` is faster 👍🏼.

|                                   Method |       Mean |     Error |    StdDev |
|----------------------------------------- |-----------:|----------:|----------:|
| SetAttributes_SetFileInformationByHandle |   151.8 ns |   2.52 ns |   2.11 ns |
|         SetAttributes_SetFileAttributesW | 9,786.4 ns | 134.68 ns | 119.39 ns |

Source: https://gist.github.com/deeprobin/2b27e68389e8d12169ac5086a8e7a00a

Reply:
If you have the handle, it's faster. You're not timing that part..

Reply:
> If you have the handle, it's faster. You're not timing that part..

@danmoseley I think the null check for SafeFileHandle.Path does not make a delay of 9,634.6ms.

Reply:
> You could use SetFileAttributesW when fileHandle.Path is not null but only if SetFileAttributesW shows better performance than SetFileInformationByHandle i.e: as an optimization.

@deeprobin you need to compare `SetFileAttributesW` vs `OpenHandle`+`SetFileInformationByHandle`+`CloseHandle`.
In your current benchmark, you've not accounting for opening and closing the handle.

Reply:
> SetFileAttributesW vs OpenHandle+SetFileInformationByHandle+CloseHandle.

These may behave different for paths that are symbolic links, cfr https://github.com/dotnet/runtime/pull/60507#discussion_r770537586.

Reply:
> 

I updated the benchmark and got other results 😄

https://gist.github.com/deeprobin/67500fdbfda1ae7f9e71277edbe9c26e

|                                   Method |      Mean |     Error |    StdDev |
|----------------------------------------- |----------:|----------:|----------:|
| SetAttributes_SetFileInformationByHandle | 18.466 us | 0.3671 us | 0.4080 us |
|         SetAttributes_SetFileAttributesW |  9.558 us | 0.1887 us | 0.2246 us |

Reply:
But in our case is `SafeFileHandle` already opened, so I think using just `SetFileInformationByHandle` is the best choice.